lib/client.js

```diff
@@ -81,16 +128,224 @@ module.exports = function (dependencies) {
+          }
+          const updatedError = { ...subDirectoryInformation, error };
+          throw updatedError;
+        }
+      }
+
+      try {
+        const sentRequest = await this.request(
+          this.manageChannelsSession,
+          this.config.manageChannelsAddress,
+          notification,
+          path,
+          httpMethod
+        );
+
+        return { ...subDirectoryInformation, ...sentRequest };
+      } catch (error) {
+        if (
+          typeof error.error !== 'undefined' &&
+          error.error.message.includes('ExpiredProviderToken')
+        ) {
+          const resentRequest = await this.retryRequest(
+            this.session,
+            this.config.address,
+            notification,
+            path,
+            httpMethod,
+            retryCount
+          );
+          return { ...subDirectoryInformation, ...resentRequest };
+        } else {
+          throw { ...subDirectoryInformation, ...error };
+        }
+      }
+    } else {
+      // Connect to standard session
+      if (!this.session || this.session.closed || this.session.destroyed) {
+        try {
+          await this.connect();
+        } catch (error) {
+          if (this.errorLogger.enabled) {
+            // Proxy server that returned error doesn't have access to logger.
+            this.errorLogger(error.message);
+          }
+          const updatedError = { ...subDirectoryInformation, error };
+          throw updatedError;
+        }
+      }
+
+      try {
+        const sentRequest = await this.request(
+          this.session,
+          this.config.address,
+          notification,
+          path,
+          httpMethod
+        );
+        return { ...subDirectoryInformation, ...sentRequest };
+      } catch (error) {
+        if (
+          typeof error.error !== 'undefined' &&
+          error.error.message.includes('ExpiredProviderToken')
+        ) {
+          const resentRequest = await this.retryRequest(
+            this.session,
+            this.config.address,
+            notification,
+            path,
+            httpMethod,
+            retryCount
+          );
+          return { ...subDirectoryInformation, ...resentRequest };
+        } else {
+          throw { ...subDirectoryInformation, ...error };
+        }
+      }
+    }
+  };
+
+  Client.prototype.retryRequest = async function retryRequest(
+    session,
+    address,
+    notification,
+    path,
+    httpMethod,
+    count
+  ) {
     if (this.isDestroyed) {
-      return Promise.resolve({ device, error: new VError('client is destroyed') });
+      const error = { error: new VError('client is destroyed') };
+      throw error;
     }
 
-    // Connect session
-    if (!this.session || this.session.closed || this.session.destroyed) {
-      return this.connect().then(() => this.request(notification, device, count));
+    const retryCount = count + 1;
+
+    if (retryCount >= this.config.connectionRetryLimit) {
```

Comment:
Closes #146 Move all retry calculations here, before, the calculation was only being used for `403` and `ExpiredProviderToken` errors and only tried a hardcoded `2` times and didn't consider this configuration value.

Now, it considers the `this.config.connectionRetryLimit` (changed default to 2 to keep original behavior) but can also be called whenever we want to add other retry mechanisms in the client.